### PR TITLE
add Pressure diag

### PR DESCRIPTION
**Pull request by nicolasaunai**
#752

Computing the pressure is like using the ParticleToMesh except we need to deposit neither 1 nor the velocity, but the velocity squared. The idea thus is to generalize the ParticleToMesh operator so, per particle, to deposit only 1 field at a time, given the field, the particle and a functor that, given the particle would:

for the density, return 1
for the flux, return vx, or vy or vz
for the pressure, return vxvx, vxvy etc.
This is done so that one day if needed to compute something else, like the heat flux (v^3) we can simply add more functors.

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `TensorField` class template to represent and manipulate tensor fields.
  - Added search functionality with a new search bar component for improved user experience.
  - Expanded the set of predefined constants for scalar quantities to enhance data representation.

- **Enhancements**
  - Upgraded the `VecField` class to utilize `TensorField`, streamlining the codebase.
  - Enhanced the `IonPopulation` and `Ions` classes with momentum tensor capabilities.
  - Improved the `Interpolator` and `ParticleToMesh` classes for more generic and versatile operations.

- **Bug Fixes**
  - Fixed issues related to momentum tensor calculations in various diagnostic classes.
  - Addressed inconsistencies in diagnostic data collection and representation.

- **Documentation**
  - Updated documentation to reflect new features and changes in diagnostic data handling.

- **Refactor**
  - Refactored the `ElectromagDiagnosticWriter` class for more efficient tensor field data handling.
  - Codebase cleanup by removing redundant includes and updating dependencies.

- **Tests**
  - Added and updated tests to cover new tensor field functionalities and momentum tensor calculations.

- **Chores**
  - Modified build configurations to enable development mode for improved developer experience.

- **Style**
  - Updated code styling for consistency across the project.

- **Revert**
  - No reverts in this release.

**Note**: The above release notes have been crafted to focus on end-user impact and do not reveal internal code structures or file names.

### CODE REVIEW

File: pyphare/pyphare/pharein/__init__.py [hunk 1]
```diff
@@ -11,14 +10,20 @@
 #    which is called from Pybind when phare-exe (or similar) is in use
 PHARE_EXE = False
 
-venv_path = os.environ.get('VIRTUAL_ENV')
+venv_path = os.environ.get("VIRTUAL_ENV")
 
 if venv_path is not None:
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharein/__init__.py:15]: @nicolasaunai do you use this anymore?

File: pyphare/pyphare/pharein/__init__.py [hunk 2]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharein/__init__.py:1]: maybe we can black these files on a standalone PR?

**Comment by nicolasaunai** [comment]
image
image

File: pyphare/pyphare/pharein/__init__.py [hunk 3]
```diff
@@ -253,7 +253,7 @@ def as_paths(rb):
 
     #### adding diagnostics
     diag_path = "simulation/diagnostics/"
-    for diag in simulation.diagnostics:
+    for _, diag in simulation.diagnostics.items():
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharein/__init__.py:256]: `for diag in simulation.diagnostics.values():`

?

might need `for diag in list(simulation.diagnostics.values()):`

File: pyphare/pyphare/pharein/diagnostics.py [hunk 4]
```diff
@@ -127,18 +126,30 @@ def __init__(self, name, **kwargs):
...
+
+        # if a diag already is registered we just contactenate the timestamps
+        addIt = True
+        registered_diags = global_vars.sim.diagnostics
+        for diagname, diag in registered_diags.items():
+            if self.quantity == diag.quantity:
+                print(
+                    f"{diag.name} already registered {self.quantity}, merging timestamps"
+                )
+                my_times = self.write_timestamps
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharein/diagnostics.py:139]: `my` -> `new`?

File: pyphare/pyphare/pharesee/run.py [hunk 5]
```diff
@@ -63,34 +64,35 @@ def _current2d(bx, by, bz, dx, dy):
...
+    elif ndim == 2:
+        Bx = patchdatas["Bx"].dataset[:]
+        By = patchdatas["By"].dataset[:]
+        Bz = patchdatas["Bz"].dataset[:]
 
-        dx, dy = patch.dl
+        dx, dy = reference_pd.dl
 
         Jx, Jy, Jz = _current2d(Bx, By, Bz, dx, dy)
-        # return ({"name":"Jx", "data":Jx,"centering":"primal"},
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/run.py:86]: torm?

File: pyphare/pyphare/pharesee/run.py [hunk 6]
```diff
@@ -108,15 +110,19 @@ def _divB2D(Bx, By, xBx, yBy):
     return dxbx + dyby
 
 
-def _compute_divB(patch):
-    if patch.box.ndim == 1:
+def _compute_divB(patchdatas, **kwargs):
+
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/run.py:114]: break here implies this file has not been formatted by black?

File: src/amr/data/field/coarsening/field_coarsen_operator.hpp [hunk 7]
```diff
@@ -71,8 +71,7 @@ namespace amr
          * In our case, we allow a RF up to 10, so having 5 ghost width is sufficient
          *
          */
-        NO_DISCARD SAMRAI::hier::IntVector
-        getStencilWidth(SAMRAI::tbox::Dimension const& dim) const override
+        SAMRAI::hier::IntVector getStencilWidth(SAMRAI::tbox::Dimension const& dim) const override
```
Comment (PhilipDeegan) [inline src/amr/data/field/coarsening/field_coarsen_operator.hpp:74]: cosmetic: could return type could be `auto`

File: src/core/data/ions/ion_population/ion_population.hpp [hunk 8]
```diff
@@ -187,10 +189,11 @@ namespace core
 
 
         NO_DISCARD VecField const& flux() const { return flux_; }
-
-
         NO_DISCARD VecField& flux() { return flux_; }
 
+        TensorField const& momentumTensor() const { return momentumTensor_; }
```
Comment (PhilipDeegan) [inline src/core/data/ions/ion_population/ion_population.hpp:194]: `NO_DISCARD` ? :)

File: src/core/data/ions/ions.hpp [hunk 9]
```diff
@@ -71,6 +73,23 @@ namespace core
                 throw std::runtime_error("Error - cannot access density data");
         }
 
+        NO_DISCARD field_type const& massDensity() const
+        {
+            if (isUsable())
+                return sameMasses_ ? *rho_ : *massDensity_;
+            else
```
Comment (PhilipDeegan) [inline src/core/data/ions/ions.hpp:80]: else here is redundant, but :shrug:

File: tests/simulator/test_diagnostics.py [hunk 10]
```diff
@@ -233,11 +233,11 @@ def _test_dump_diags(self, dim, **simInput):
             self.simulator = None
             ph.global_vars.sim = None
 
-    def test_twice_register(self):
-        simulation = ph.Simulation(**simArgs.copy())
-        model = setup_model()
-        dump_all_diags(model.populations)  # first register
-        self.assertRaises(RuntimeError, dump_all_diags, model.populations)
+    # def test_twice_register(self):
```
Comment (PhilipDeegan) [inline tests/simulator/test_diagnostics.py:236]: ah?

File: src/restarts/detail/h5writer.hpp [hunk 11]
```diff
@@ -1,12 +1,10 @@
...
 #ifndef PHARE_DETAIL_RESTART_HIGHFIVE_HPP
 #define PHARE_DETAIL_RESTART_HIGHFIVE_HPP
 
-#include "core/logger.hpp"
-#include "core/utilities/types.hpp"
 
+#include "core/utilities/mpi_utils.hpp"
 #include "restarts/restarts_props.hpp"
-
+#include "initializer/data_provider.hpp"
```
Comment (PhilipDeegan) [inline src/restarts/detail/h5writer.hpp:7]: I don't get it

File: src/diagnostic/detail/types/fluid.hpp [hunk 12]
```diff
@@ -98,28 +184,41 @@ void FluidDiagnosticWriter<H5Writer>::getDataSetInfo(DiagnosticProperties& diagn
     };
 
     auto infoVF = [&](auto& vecF, std::string name, auto& attr) {
-        for (auto const& [id, type] : core::Components::componentMap)
+        for (auto const& [id, type] : core::Components::componentMap<1>())
```
Comment (PhilipDeegan) [inline src/diagnostic/detail/types/fluid.hpp:187]: kinda magic

possibly https://github.com/PhilipDeegan/PHARE/pull/83/commits/fb1bb5eef6c86ffadc4f1276f3d4a029cc807f6f#diff-064aa83f5fab23f8a384cc61d69e5847c788b7aa257b3f905410149091588d1eR56 ?

File: src/diagnostic/detail/types/electromag.hpp [hunk 13]
```diff
@@ -104,7 +104,7 @@
     auto vecFields = h5Writer.modelView().getElectromagFields();
 
     auto initVF = [&](auto& path, auto& attr, std::string key, auto null) {
```
[the remark this answers is not in the document]
Comment (PhilipDeegan) [reply src/diagnostic/detail/types/electromag.hpp:106]: :face_in_clouds:

File: src/core/data/grid/gridlayoutimplyee.hpp [hunk 14]
```diff
@@ -91,11 +91,25 @@ namespace core
             const std::array<QtyCentering, NBR_COMPO> Vz
                 = {{data.primal, data.primal, data.primal}};
 
+            const std::array<QtyCentering, NBR_COMPO> Mxx
```
Comment (PhilipDeegan) [inline src/core/data/grid/gridlayoutimplyee.hpp:94]: `clang-format-14` has an option to move consts to the right

```
QualifierAlignment: Right
```

I think

File: pyphare/pyphare/pharein/diagnostics.py (hunk 4 above)
Comment (nicolasaunai) [reply pyphare/pyphare/pharein/diagnostics.py:139]: I think the proposed change is misleading. 'my' refers to 'self', 'new' refers to future self.
adding self.write_timestamps in the np.concatenates makes the line quite long and I find

' new_times = np.concatenate((my_times, existing_times))` 

quite easy to read.

I don't think there is a perf issue as everything is just references anyway.

File: pyphare/pyphare/pharesee/run.py (hunk 6 above)
Comment (nicolasaunai) [reply pyphare/pyphare/pharesee/run.py:114]: not sure what you mean I have black on save and this is version 22.8

File: pyphare/pyphare/pharein/diagnostics.py (hunk 4 above)
Comment (PhilipDeegan) [reply pyphare/pyphare/pharein/diagnostics.py:139]: `new` IMO refers to additional, if there is existing

I have an aversion to `my`, implies personhood

File: pyphare/pyphare/pharesee/run.py (hunk 6 above)
Comment (PhilipDeegan) [reply pyphare/pyphare/pharesee/run.py:114]: doesn't look like it

File: src/restarts/detail/h5writer.hpp (hunk 11 above)
Comment (nicolasaunai) [reply src/restarts/detail/h5writer.hpp:7]: you don't get what?

Comment (nicolasaunai) [reply src/restarts/detail/h5writer.hpp:7]: data provider adds PhareDict

Comment (PhilipDeegan) [reply src/restarts/detail/h5writer.hpp:7]: it was implicit before I guess, cause if it was needed it wouldn't compile without this,

just seems a bit odd on it's own without other edits in this file

**Review by PhilipDeegan — APPROVED** [review]

**Pull request merged**